service/controller/v12/resource/instance/checksum.go

```diff
@@ -0,0 +1,99 @@
+package instance
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2018-05-01/resources"
+	"io/ioutil"
+	"net/http"
+)
+
+func getDeploymentTemplateChecksum(deployment resources.Deployment) (string, error) {
+
+	templateLink := deployment.Properties.TemplateLink
+	if templateLink == nil {
+		return "", nilTemplateLinkError
+	}
+
+	resp, err := http.Get(*templateLink.URI)
```

Comment:
Not sure I got your point here. When would this case happen?
The checksum is calculated twice:
1) whenever sending the updated deployment
2) when checking if the template has changed

I don't see in which case we could avoid making the calculation.

Reply:
There are few assumptions here so bear with me:

When the template already exists as deployment, I'd expect something like following if I have understood everything right:

*CR changes / upgrade changes parameters*:
1. Compute checksum for template and its parameters.
2. Compare checksum against the one in CR status.
3. If it doesn't match, update deployment. No need to check the deployment in Azure.

*CR doesn't change / reconcile Azure state*
1. Compute checksum for template and its parameters.
2. Compare checksum against the one in CR status.
3. It matches.
4. Invoke conditional GET with checksum against Azure to see if template or parameters in there are up to date.
5. If either of them is not up to date, `CreateOrUpdate` latest deployment.

Does this make sense? Or am I somehow misunderstanding something?

Reply:
It would make sense in theory, but you are misunderstanding one thing I guess.
The checksum is never made towards the Azure deployment. that's simply not possible.
When the azure deployment is updated, the checksum is calculated and stored in the CR. From that moment on, we assume that what's in ARM has the checksum that we saved in the CR.
If for any reason the azure deployment changes because of an external factor, the checksum stored in the CR would be simply wrong.

The HTTP get call we make is towards github to download the actual ARM template file content:

> https://raw.githubusercontent.com/giantswarm/azure-operator/%s/service/controller/v12/resource/%s/template/%s

Reply:
Ah, right. Makes sense. All good here.